### PR TITLE
OPENEUROPA-2041: Change url of textarea help link.

### PR DESCRIPTION
**Pull request by imanoleguskiza**
### Description

Change url of textarea help link to point to the actual text format help.

### Change log

- Added:
- Changed: Change url of textarea help link to point to the actual text format help.
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: tests/features/textarea_help.feature [hunk 1]
```diff
@@ -0,0 +1,10 @@
+@ap
```
Comment (nagyad) [inline tests/features/textarea_help.feature]: typo

File: tests/features/textarea_help.feature [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+Feature: Textarea help
+  As a content editor
+  When I edit content
+  The textarea help should take me to the relevant help page
```
Comment (nagyad) [inline tests/features/textarea_help.feature]: The textarea help should take me to the relevant text format help page

File: tests/features/textarea_help.feature [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+  Scenario: Textarea help links point to the proper help page.
+    Given I am logged in as a user with the "Author" role
+    When I visit "the demo content creation page"
+    And I should see the link "About text formats" point to "the plain text help page"
```
Comment (nagyad) [inline tests/features/textarea_help.feature]: Empty line in the end of file.

File: oe_editorial.module [hunk 4]
```diff
@@ -6,3 +6,26 @@
...
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Url;
+
+/**
+ * Implements hook_field_widget_form_alter().
+ */
+function oe_editorial_field_widget_form_alter(&$element, FormStateInterface $form_state, $context) {
+  // We add an after build method so we can alter the Url of the help link.
+  if ($element['#type'] == 'text_format') {
```
Comment (nagyad) [inline oe_editorial.module]: ===

File: modules/oe_editorial_workflow_demo/config/install/core.entity_form_display.node.oe_workflow_demo.default.yml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  uid:
+    type: entity_reference_autocomplete
+    weight: 2
+    settings:
+      match_operator: CONTAINS
+      size: 60
+      placeholder: ''
+    region: content
+    third_party_settings: {  }
+hidden: {  }
```
Comment (nagyad) [inline modules/oe_editorial_workflow_demo/config/install/core.entity_form_display.node.oe_workflow_demo.default.yml]: Empty line.

File: runner.yml.dist [hunk 6]
```diff
@@ -20,6 +20,7 @@ drupal:
     - "./vendor/bin/drush config-set node.settings use_admin_theme 1 -y"
     # Enable the modules.
     - "./vendor/bin/drush en config_devel -y"
+    - "./vendor/bin/drush en oe_editorial -y"
```
Comment (nagyad) [inline runner.yml.dist]: Why is this required?

Comment (imanoleguskiza) [reply runner.yml.dist]: oe_editorial is not a requirement of oe_editorial_demo so it doesn't get installed by default. Do you think we should add it as a dependency?

Comment (nagyad) [reply runner.yml.dist]: Yeah right, if you go through the chain, oe_editorial_demo->oe_editorial_entity_version->oe_editorial_corporate_workflow->oe_editorial, so please add oe_editorial as dependency to oe_editorial_corporate_workflow instead of having it in the runner.

File: oe_editorial.module [hunk 7]
```diff
@@ -6,3 +6,50 @@
...
+  ])) {
+    $element['#after_build'][] = '_oe_editorial_alter_textarea_help';
+  }
+}
+
+/**
+ * After build callback to alter the Url of the help link on text areas.
+ */
+function _oe_editorial_alter_textarea_help($form_element, FormStateInterface $form_state) {
+  if (isset($form_element['format']['format'])) {
```
Comment (nagyad) [inline oe_editorial.module]: Could we start this as: 
if (!isset($form_element['format']['format'])) {
  return;
}
...

File: tests/Behat/TextFormatContext.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+namespace Drupal\Tests\oe_editorial\Behat;
+
+use Behat\Behat\Hook\Scope\AfterScenarioScope;
+use Drupal\DrupalExtension\Context\RawDrupalContext;
+use Behat\Gherkin\Node\TableNode;
+use Drupal\filter\Entity\FilterFormat;
+use Drupal\user\Entity\Role;
+
+/**
+ * Defines step definitions for testing the Content Lock feature.
```
Comment (nagyad) [inline tests/Behat/TextFormatContext.php]: Copy paste. Please fix the feature name mentioned.

File: tests/Behat/TextFormatContext.php [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+ */
+class TextFormatContext extends RawDrupalContext {
+
+  /**
+   * The IDs of the existing text formats.
+   *
+   * @var array
+   */
+  protected $newTextFormats = [];
```
Comment (nagyad) [inline tests/Behat/TextFormatContext.php]: Are these existing, or new text formats?

File: tests/Behat/TextFormatContext.php [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+  /**
+   * Deletes text formats created through the scenario..
```
Comment (nagyad) [inline tests/Behat/TextFormatContext.php]: .. in the end!

File: oe_editorial.libraries.yml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+help:
```
Comment (sergepavle) [inline oe_editorial.libraries.yml]: I think that that this js file in fact modification of drupal.filter library's js we have to use hook alter for libraries.

File: oe_editorial.module [hunk 12]
```diff
@@ -6,3 +6,52 @@
...
+          'filter-help-item',
+          'filter-help-' . $allowed_format_id,
+        ],
+        'target' => '_blank',
+      ],
+    ];
+  }
+  // Remove the default static link.
+  unset($form_element['format']['help']['about']);
+  $form_element['format']['help']['#attached']['library'][] = 'oe_editorial/help';
```
Comment (sergepavle) [inline oe_editorial.module]: If possible, I think that it could be better to use altering of original library.

File: tests/Behat/FeatureContext.php [hunk 13]
```diff
@@ -199,4 +199,33 @@ public function theNodeShouldHaveTheFollowingVersion(string $title, TableNode $o
...
+   *
+   * @param string $link
+   *   Link identifier.
+   * @param string $path
+   *   Target path of the link.
+   *
+   * @throws \Exception
+   *   Throws an exception if the link is not found or if the target is wrong.
+   *
+   * @Then I should see the link :link point to :path
```
Comment (sergepavle) [inline tests/Behat/FeatureContext.php]: Is this possible to reuse this function inside oe_content lib
\Drupal\Tests\oe_content\Behat\FeatureContext::assertLinkWithHref
?

File: tests/Behat/TextFormatContext.php [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+      }
+      $text_format = FilterFormat::load($text_format_id);
+      $text_format->delete();
+    }
+  }
+
+  /**
+   * Checks the given node with title has the specified version numbers.
+   *
+   * // phpcs:disable
```
Comment (sergepavle) [inline tests/Behat/TextFormatContext.php]: If you have reasonable reason for disabling CS for this part of code, probably better to disable specific functionality of code sniffer.

File: oe_editorial.libraries.yml (hunk 11 above)
Comment (brummbar) [reply oe_editorial.libraries.yml]: We discussed with @sergepavle and we are adding extra js functionality so we can't extend the original.

File: tests/Behat/FeatureContext.php (hunk 13 above)
Comment (brummbar) [reply tests/Behat/FeatureContext.php]: We can, let's see if it doesn't clash for other tests.

File: tests/Behat/TextFormatContext.php (hunk 14 above)
Comment (brummbar) [reply tests/Behat/TextFormatContext.php]: The @Then should not have any examples.
If you want to put an example, put it just under the "Checks the given node..." and yes it needs to be wrapped in full phpcs:disable like now.

File: oe_editorial.module [hunk 15]
```diff
@@ -6,3 +6,52 @@
...
+
+/**
+ * Implements hook_field_widget_form_alter().
+ */
+function oe_editorial_field_widget_form_alter(&$element, FormStateInterface $form_state, $context) {
+  /** @var \Drupal\Core\Field\FieldDefinitionInterface $field_definition */
+  $field_definition = $context['items']->getFieldDefinition();
+
+  // We add an after build method so we can alter the Url of the help link.
+  if (in_array($field_definition->getType(), [
```
Comment (brummbar) [inline oe_editorial.module]: Let's try to target all the possible "text_format" elements by adding a process function there.

File: tests/features/text_format_help.feature [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+  @javascript @cleanup-formats
+  Scenario: Text format help links point to the proper help page.
+    Given I am logged in as a user with the "Author" role
+    And the following text formats are available:
+      | id               | title              |
+      | rich_text        | Rich text          |
+      | simple_rich_text | Simple rich text   |
+    When I visit "the demo content creation page"
+    Then I should see "Rich Text"
+    And I should see the link "About the Rich text format" point to "the rich text help page"
```
Comment (brummbar) [inline tests/features/text_format_help.feature]: Let's add a test that checks that the guidelines are hidden now.

File: tests/Behat/TextFormatContext.php (hunk 14 above)
Comment (brummbar) [reply tests/Behat/TextFormatContext.php]: This needs to be fixed.

File: tests/Behat/FeatureContext.php (hunk 13 above)
Comment (brummbar) [reply tests/Behat/FeatureContext.php]: This too.

Comment (imanoleguskiza) [reply tests/Behat/FeatureContext.php]: But oe_content doesn't have any direct dependency with oe_editorial (nor do we have it the other way around), so how am I supposed to use this function?

File: tests/Behat/TextFormatContext.php (hunk 14 above)
Comment (imanoleguskiza) [reply tests/Behat/TextFormatContext.php]: Fixed

File: tests/features/text_format_help.feature (hunk 16 above)
Comment (imanoleguskiza) [reply tests/features/text_format_help.feature]: Added

File: tests/Behat/FeatureContext.php (hunk 13 above)
Comment (sergepavle) [reply tests/Behat/FeatureContext.php]: Yes I agree doubts of Imanol. Don't sure that it make sense add additional dependencies from oe_content.
